### PR TITLE
Enable smart weight loading 

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
Enable smart weight loading for all task.

### **Note**
**There are some exceptions in copying algorithms for two stage detectors, and ssd-mobilenetv2
They are resolved by next PR**

### How to test
I added unit tests for smart weight loading algo for general cases

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by vinnamkim — COMMENTED** [review]
I have questions:
1. What is the smart weight loading?
2. Which does an user scenario this smart weight loading cover?

**Comment by jaegukhyun** [comment]
> I have questions:
> 
> 1. What is the smart weight loading?
> 2. Which does an user scenario this smart weight loading cover?

Smart weight loading is weight loading method for class incremental learning. 
When user trained their model on dataset whose classes are ["car", "bus"]
And user wants to train their trained model with the new dataset whose classes are ["car", "bus", "truck"]
In this case, without smart weight loading, the final classification layer will be initialized since dimension of final classification layer between checkpoint and model are different.
Here smart weight loading will load weights from checkpoint for previous classes and initialize weights for newly added classes.

**Comment by vinnamkim** [comment]
> Smart weight loading is weight loading method for class incremental learning. When user trained their model on dataset whose classes are ["car", "bus"] And user wants to train their trained model with the new dataset whose classes are ["car", "bus", "truck"] 

It seems that we should add a dataset-level meta info to `OTXModel`. Let's call it as `PredictionMetaInfo` (you can name a more appropriate one for it, shorter one) such as

```python
class PredictionMetaInfo:
     pass

@dataclasses
class MultiClassClsPredictionMetaInfo(PredictionMetaInfo):
     num_classes: int
     class_names: list[str]
```

Then, assume `OTXModel` should be created by `OTXModel(prediction_meta_info=PredictionMetaInfo)`.

Now, you can handle weight loading in class incremental learning scenarios by

```python
ckpt = old_model.state_dict()

assert "prediction_meta_info" in ckpt

new_model = OTXModel(prediction_meta_info=new_prediction_meta_info)

new_model.load_state_dict(ckpt) # Smart weight loading can be done

class OTXModel(nn.Module):
     ...

     def load_state_dict(self, ckpt):
           old_info = ckpt["prediction_meta_info"]
           new_info = self.prediction_meta_info
           
           if old_info != new_info:
               self._load_state_dict_incrementally(ckpt)
           else:
               super().load_state_dict(ckpt)

     def _load_state_dict_incrementally(self, ckpt):
           ...

     def state_dict(self):
           ckpt = super().state_dict()
           ckpt["prediction_meta_info"] = asdict(self.prediction_meta_info)
```

This requirement is not only observed in this PR, but also @sungmanc's today presentation,
image
This `HLabelInfo` includes the dataset-level meta information, but is passed in each data sample. In addition, it is related to ModelAPI, so that it is not just required for H-label but for all other tasks.


p.s. Is it fundamentally possible to solve the following problem by implementing own `_load_state_dict_incrementally()` for each `OTXModel`? We can have some general solutions at some model classes, but it is not possible to infer what is the head layer of the model by machine automatically.
> There are some exceptions in copying algorithms for two stage detectors, and ssd-mobilenetv2. They are resolved by next PR

**Comment by sungmanc** [comment]
> It seems that we should add a dataset-level meta info to `OTXModel`. Let's call it as `PredictionMetaInfo` (you can name a more appropriate one for it, shorter one)

Sounds good to me, if we can feed a dataset-lele meta info to the `ModelEntity`, it's great

**Comment by jaegukhyun** [comment]
@vinnamkim @sungmanc 
I'll update codes from updating meta info from dataset to updating meta info from model

File: src/otx/core/engine/train.py [hunk 1]
```diff
@@ -108,7 +111,15 @@ def train(
...
-        trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)
+        if cfg.resume:
+            trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)
+        else:
+            # load weight
+            if cfg.checkpoint is not None:
+                loaded_checkpoint = torch.load(cfg.checkpoint)
+                model.load_state_dict(loaded_checkpoint["state_dict"])
+            # train
+            trainer.fit(model=model, datamodule=datamodule)
```
Comment (vinnamkim) [inline src/otx/core/engine/train.py]:
```suggestion
       # load weight
       if cfg.checkpoint is not None:
           loaded_checkpoint = torch.load(cfg.checkpoint)
           model.load_state_dict(loaded_checkpoint["state_dict"])
       # train
       trainer.fit(model=model, datamodule=datamodule)
```

Is it impossible to drop `cfg.resume`?

File: src/otx/core/engine/train.py [hunk 2]
```diff
@@ -69,9 +70,11 @@ def train(
 
     log.info(f"Instantiating datamodule <{cfg.data}>")
     datamodule = OTXDataModule(task=cfg.base.task, config=cfg.data)
+    data_meta_info = datamodule.meta_info
 
     log.info(f"Instantiating model <{cfg.model}>")
     model: OTXLitModule = hydra.utils.instantiate(cfg.model)
+    model.meta_info = data_meta_info
```
Comment (sungmanc) [inline src/otx/core/engine/train.py]:
```suggestion
    log.info(f"Instantiating model <{cfg.model}>")
    model: OTXLitModule = hydra.utils.instantiate(cfg.model)
    model.meta_info = datamodule.meta_info
```

File: src/otx/core/engine/train.py [hunk 3]
```diff
@@ -108,7 +111,15 @@ def train(
 
     if cfg.train:
         log.info("Starting training!")
-        trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)
+        if cfg.resume:
+            trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)
+        else:
+            # load weight
```
Comment (sungmanc) [inline src/otx/core/engine/train.py]:
```suggestion
            # load weight
```
```suggestion
            # load weight to finetune the model
```

File: src/otx/core/model/module/base.py [hunk 4]
```diff
@@ -105,7 +108,58 @@ def configure_optimizers(self) -> dict[str, Any]:
...
+        return state_dict
+
+    def load_state_dict(self, state_dict: dict[str, Any]) -> None:
+        """Load state dictionary from checkpoint state dictionary.
+
+        If checkpoint's meta_info and OTXLitModule's meta_info are different,
+        load_state_pre_hook for smart weight loading will be registered.
+        """
+        ckpt_meta_info = state_dict.pop("meta_info", None)
+        if ckpt_meta_info and self.meta_info and ckpt_meta_info != self.meta_info:
```
Comment (vinnamkim) [inline src/otx/core/model/module/base.py:145]:
```suggestion
        # ckpt_meta_info exists but self.meta_info does not exists
        if ckpt_meta_info and self.meta_info is None:
            msg = "`state_dict` to load has `meta_info`, but the current model has no `meta_info`. " \ 
                       "It is recommended to set proper `meta_info` for the incremental learning case."
            warn(msg)
            return super().load_state_dict(state_dict)
        # This is the case
        if ckpt_meta_info and self.meta_info and ckpt_meta_info == self.meta_info:
            ...
            return ...
        return super().load_state_dict()
```
Please add an user warning for there is `meta_info` in `state_dict` but not in `self`. In addition, there is a return object in `load_state_dict()`. Please see https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.load_state_dict

**Review by sungmanc — DISMISSED** [review]

File: src/otx/core/data/module.py [hunk 5]
```diff
@@ -24,6 +26,21 @@
     from .dataset.base import OTXDataset
 
 
+@dataclass
+class DataMetaInfo:
```
Comment (sungmanc) [inline src/otx/core/data/module.py:30]: It seems that we need to consider the `BaseClass` (maybe `BaseDataMeta`?) to broaden the tasks. What do you think ? In my case, I need to add additional `HLabelInfo` dataclass and it would be better if there is a baseclass to inherit.

File: src/otx/core/model/entity/base.py [hunk 6]
```diff
@@ -57,3 +60,49 @@ def forward(
...
+        self._register_load_state_dict_pre_hook(self.load_state_dict_pre_hook)
+
+    def load_state_dict_pre_hook(self, state_dict: dict[str, torch.Tensor], prefix: str, *args, **kwargs) -> None:
+        """Modify input state_dict according to class name matching before weight loading."""
+        model2ckpt = self.map_class_names(self.model_classes, self.ckpt_classes)
+
+        for param_name in self.state_dict():
+            model_param = self.state_dict()[param_name].clone()
+            ckpt_param = state_dict[prefix + param_name]
+            if model_param.shape != ckpt_param.shape:
```
Comment (vinnamkim) [inline src/otx/core/model/entity/base.py]: How can we cover the following case?

```python
          src_classes = ['person', 'car', 'tree']
          dst_classes = ['sky', 'ball', 'person']
```

Comment (vinnamkim) [reply src/otx/core/model/entity/base.py]: I believe that this can be fundamentally solved by implementing a logic to search the prediction layer for each `OTXModel`. Like

> p.s. Is it fundamentally possible to solve the following problem by implementing own _load_state_dict_incrementally() for each OTXModel? We can have some general solutions at some model classes, but it is not possible to infer what is the head layer of the model by machine automatically.

Comment (vinnamkim) [reply src/otx/core/model/entity/base.py]: I believe that `meta_info` decides the shape of `OTXModel`, so that `meta_info` should be a member of `OTXModel`. In addition, it means that `OTXModel.load_state_dict()` should be able to handle the increamental learning case by itself.

File: src/otx/core/data/module.py (hunk 5 above)
Comment (jaegukhyun) [reply src/otx/core/data/module.py:30]: How about HLabelDataMetaInfo(DataMetaInfo)?

Comment (vinnamkim) [reply src/otx/core/data/module.py:30]: I told @jaegukhyun to share `DataMetaInfo` across all tasks. `HLabelInfo` can be inherited from `DataMetaInfo`. However, `DataMetaInfo` seems not appealing name to me. We might have a better name in the future or this time by someone.

File: src/otx/core/engine/train.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/engine/train.py]: Furthermore, if `model` can handle `cfg.checkpoint` having `meta_info`, do we need this change?

Comment (jaegukhyun) [reply src/otx/core/engine/train.py]: Lightening trainer will load other module's hyperparameters such as optimization, lr scheduling when `trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)`
This is useful feature when user wants resume, but this could be unwanted behavior those who wants retrain with previous checkpoint

File: src/otx/core/model/module/base.py (hunk 4 above)
Comment (jaegukhyun) [reply src/otx/core/model/module/base.py:145]: Done by 5bfff3084cfdc99a987f9d3b516c488219b48ca9

File: src/otx/core/model/entity/base.py (hunk 6 above)
Comment (jaegukhyun) [reply src/otx/core/model/entity/base.py]: Done by 4517f09 and 486a2e3
I added auto detection of classification layers by comparing two models with different num_classes

File: src/otx/core/model/entity/action_classification.py [hunk 7]
```diff
@@ -13,7 +13,7 @@
 )
 from otx.core.data.entity.base import OTXBatchLossEntity
 from otx.core.model.entity.base import OTXModel
-from otx.core.utils.build import build_mm_model
+from otx.core.utils.build import build_mm_model, get_classification_layers
```
Comment (sungmanc) [inline src/otx/core/model/entity/action_classification.py:16]: I'm not sure about the naming of `get_classification_layers`, however, I also have no better idea. I hope we could consider the better naming later.

**Review by sungmanc — DISMISSED** [review]

File: src/otx/core/model/entity/classification.py [hunk 8]
```diff
@@ -46,7 +46,8 @@ def device(self) -> device:
             else:
                 return buf.device
 
-    return build_mm_model(config, MODELS, load_from)
+    classification_layers = get_classification_layers(config, MODELS, "model.")
+    return build_mm_model(config, MODELS, load_from), classification_layers
```
Comment (vinnamkim) [inline src/otx/core/model/entity/classification.py:50]: Is this possible to return the tuple although the return type is `nn.Module` at the line 32?

File: src/otx/core/model/entity/classification.py [hunk 9]
```diff
@@ -63,7 +64,9 @@ def __init__(self, config: DictConfig) -> None:
         super().__init__()
 
     def _create_model(self) -> nn.Module:
-        return _create_mmpretrain_model(self.config, self.load_from)
+        model, classification_layers = _create_mmpretrain_model(self.config, self.load_from)
+        self.classification_layers = classification_layers
```
Comment (vinnamkim) [inline src/otx/core/model/entity/classification.py:68]: Why don't we create them line by line?
```suggestion
        model = _create_mmpretrain_model(self.config, self.load_from)
        self.classification_layers = get_classification_layers(self.config)
```

Comment (jaegukhyun) [reply src/otx/core/model/entity/classification.py:68]: get_classification_layers funciton require model registry to build model with different num classes

File: src/otx/core/model/entity/classification.py (hunk 8 above)
Comment (jaegukhyun) [reply src/otx/core/model/entity/classification.py:50]: Changed by 
b095a26 It seems mypy do not test return type strictly

**Review by vinnamkim — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by harimkang — APPROVED** [review]
